### PR TITLE
add tests for Gauge, Set

### PR DESCRIPTION
**Pull request by shintasmith**
**Why:**
The goal is to increase test coverage of the various Serialization/De-serialization code, particularly in the are of Enum and aggregated metrics.

**How:**
* add Gauge and Set serialization/de-serialization tests
* stop using serialized files in SerializationTest class
* move some String related serialization/de-serialization tests from SerializationTest.class to its own test classes and add more tests to increase coverage. Support for String metrics is no longer available in the main code base

**Not Included:**
* didn't improve coverage for Histogram serialization/de-serialization code. I was told Histogram is to be deprecated
* didn't improve coverage for SimpleNumberSerialization. I could not get the normal round trip test to work.

### CODE REVIEW

**Comment by shintasmith** [comment]
@usnavi @ChandraAddala @VinnyQ @izrik @drenalin23 please review.

cc: @tilogaat @GeorgeJahad

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/io/serializers/GaugeRollupSerializerTest.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+        bb = ByteBuffer.wrap(Base64.decodeBase64(reader.readLine().getBytes()));
+        BluefloodGaugeRollup deserializedGauge3 = NumericSerializer.serializerFor(BluefloodGaugeRollup.class).fromByteBuffer(bb);
+        Assert.assertEquals(gauge3, deserializedGauge3);
+
+        bb = ByteBuffer.wrap(Base64.decodeBase64(reader.readLine().getBytes()));
+        BluefloodGaugeRollup deserializedGauge4 = NumericSerializer.serializerFor(BluefloodGaugeRollup.class).fromByteBuffer(bb);
+        Assert.assertEquals(gaugesRollup, deserializedGauge4);
+
+        Assert.assertFalse(deserializedGauge1.equals(deserializedGauge2));
```
Comment (usnavi) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/io/serializers/GaugeRollupSerializerTest.java]: Some people might split this into 3 tests, I think @izrik prefer things that way.  We can cover this in our dev meeting.

Comment (usnavi) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/io/serializers/GaugeRollupSerializerTest.java]: nm, i'm on crack.

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/io/serializers/StringSerializerTest.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    // catch that error and assert it happens only when expected.
+    @Test (expected = UnexpectedStringSerializationException.class)
+    public void testStringFullSerializationShouldFail() throws Throwable {
+        ByteArrayOutputStream baos = serializeString(TEST_STRING);
+        ByteBuffer byteBuffer = ByteBuffer.wrap(baos.toByteArray());
+        try {
+            NumericSerializer.serializerFor(Object.class).fromByteBuffer(byteBuffer);
+        } catch (RuntimeException re) {
+            throw re.getCause();
+        }
```
Comment (usnavi) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/io/serializers/StringSerializerTest.java]: just curious, why catch to only rethrow?

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/io/serializers/StringSerializerTest.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            throw ex.getCause();
+        }
+    }
+
+    // At some point of time, we had some serializer for String class.
+    // We serialized one string metric and saved it to the file
+    // src/test/resources/serializers/full_version_0.bin.
+    // This method was constructed after analyzing the content
+    // of that file. The actual code serializing the string in
+    // the main code is no longer there.
```
Comment (usnavi) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/io/serializers/StringSerializerTest.java]: buah?!

**Comment by usnavi** [comment]
lgtm

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/io/serializers/StringSerializerTest.java (hunk 2 above)
Comment (shintasmith) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/io/serializers/StringSerializerTest.java]: The NumericSerializer.serializerFor() code wraps any UnexpectedStringSerializerException into a RuntimeException. I want to check specificly for the UnexpectedStringSerializerException, so if I'm getting a different exception, it should be considered test failure.

Comment (usnavi) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/io/serializers/StringSerializerTest.java]: gotcha.  thanks!

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/io/serializers/BaseRollupSerializerTest.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package com.rackspacecloud.blueflood.io.serializers;
+
+import com.rackspacecloud.blueflood.types.Points;
+
+/**
+ * Holds some methods shared across many RollupSerializerTest classes
+ */
+public class BaseRollupSerializerTest {
+
+    protected <T> Points<T> asPoints(Class<T> type, long initialTime, long timeDelta, T... values) {
```
Comment (ChandraAddala) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/io/serializers/BaseRollupSerializerTest.java]: should T be bounded to Rollup class?

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/types/BaseRollupTest.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.rackspacecloud.blueflood.types;
+
+/**
+ * Holds methods shared by many RollupTest classes
+ */
+public class BaseRollupTest {
+
+    protected <T> Points<T> asPoints(Class<T> type, long initialTime, long timeDelta, T... values) {
+        Points<T> points = new Points<T>();
```
Comment (ChandraAddala) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/types/BaseRollupTest.java]: This methods looks similar to the one in BaseRollupSerializerTest. Are they different?

**Comment by ChandraAddala** [comment]
Apart from my comments, it looks good.

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/types/BaseRollupTest.java (hunk 5 above)
Comment (shintasmith) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/types/BaseRollupTest.java]: They are the same exact copy. Previously, the tests in package com.rackspacecloud.blueflood.io.serializers use the BluefloodEnumRollupTest.asPoints() (the BluefloodEnumRollupTest is in com.rackspacecloud.blueflood.types). I didn't like that unnecessary dependency across packages like that. So I moved the BluefloodEnumRollupTest.asPoints() into a BaseRollupTest class. In the future, when we clean up the RollupTests, we can reuse this method.

File: blueflood-core/src/test/java/com/rackspacecloud/blueflood/utils/Rollups.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package com.rackspacecloud.blueflood.utils;
+
+import com.rackspacecloud.blueflood.types.Points;
+import com.rackspacecloud.blueflood.types.Rollup;
+
+public class Rollups {
+
+    public static <T extends Rollup> Points<T> asPoints(Class<T> type, long initialTime, long timeDelta, T... values) {
```
Comment (shintasmith) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/utils/Rollups.java]: @ChandraAddala , I moved the asPoints() method here.

Comment (ChandraAddala) [inline blueflood-core/src/test/java/com/rackspacecloud/blueflood/utils/Rollups.java]: looks good.

**Pull request merged**